### ISSUE TITLE
Single event title getting displayed multiple times

### ISSUE THREAD

**Issue by ranjithnair02** [alamkanak/Android-Week-View#81]
I am displaying the weekview in a DialogFragment. There is just one event added to it but the problem now is that the same event title is getting displayed multiple times within the event. As I see in the screenshots I believe just one title should be displayed.
My Code:

    public class WeekviewFragment extends DialogFragment implements WeekView.MonthChangeListener, WeekView.EventClickListener {
    private WeekView mWeekView;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.weekview, container, false);
        mWeekView = (WeekView) view.findViewById(R.id.weekView);
        mWeekView.setMonthChangeListener(this);
        mWeekView.setOnEventClickListener(this);
        return view;
    }

    @Override
    public List<WeekViewEvent> onMonthChange(int i, int i2) {
        // Populate the week view with some events.
        List<WeekViewEvent> events = new ArrayList<>(1);

        Calendar startTime = Calendar.getInstance();
        startTime.set(Calendar.HOUR_OF_DAY, 3);
        startTime.set(Calendar.MINUTE, 0);
        startTime.set(Calendar.MONTH, 1);
        startTime.set(Calendar.YEAR, 2015);
        Calendar endTime = (Calendar) startTime.clone();
        endTime.add(Calendar.HOUR_OF_DAY, 4);
        endTime.set(Calendar.MONTH, 1);
        WeekViewEvent event = new WeekViewEvent(1, "This is title", startTime, endTime);
        event.setColor(getResources().getColor(R.color.accent_color_dark));
        events.add(event);
        return events;
    }
}

Layout code:

    <?xml version="1.0" encoding="utf-8"?>
    <RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
                xmlns:tools="http://schemas.android.com/tools"
                xmlns:app="http://schemas.android.com/apk/res-auto"
                android:layout_width="match_parent"
                android:layout_height="match_parent"
                tools:context=".MyActivity"
                android:background="#fff">

    <com.alamkanak.weekview.WeekView
        android:id="@+id/weekView"
        android:layout_width="match_parent"
        android:layout_height="match_parent"
        app:eventTextColor="@android:color/white"
        app:textSize="12sp"
        app:hourHeight="60dp"
        app:headerColumnPadding="8dp"
        app:headerColumnTextColor="#8f000000"
        app:headerRowPadding="12dp"
        app:columnGap="8dp"
        app:noOfVisibleDays="3"
        app:headerRowBackgroundColor="#ffefefef"
        app:dayBackgroundColor="#05000000"
        app:todayBackgroundColor="#1848adff"
        app:headerColumnBackground="#ffffffff"/>
  </RelativeLayout>

![image](https://cloud.githubusercontent.com/assets/9027978/6315401/876722be-ba26-11e4-88b0-cc96c003e845.png)

**Comment by caske33** [comment]
The onMonthChange is called three times: One time for this month and then one time for the previous and next month (for smooth scrolling). Because you're providing the events three times, they are rendered three times.

(you should only return the events that fall within that month)

**Comment by dgrant** [comment]
I'm seeing this too.

**Comment by alamkanak** [comment]
@caske33 explains it correctly. Just check if the month matched with your event before returning it in `onMonthChange()`

**Comment by dgrant** [comment]
Yep, makes sense now... you feeling better now @alamkanak?

**Comment by hdcoder** [comment]
What about the date which is extended for the next month ? 
Eg 30/6/2015 to event 1/7/2015
How to go about such type ?

**Comment by Shajeel-Afzal** [comment]
@caske33 

I don't think that the onMonthChange is called this way, i think it is being called as 1- Previous month, 2- current month, 3- Next month.

**Comment by Swapnil-Garg16** [comment]
There is no such mistake as discussed above in my code still it is displaying the event 3 times 

package com.example.hp.client;

import android.content.DialogInterface;
import android.graphics.RectF;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import com.alamkanak.weekview.WeekView;
import com.alamkanak.weekview.WeekViewEvent;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;

public class MainActivity extends AppCompatActivity implements WeekView.MonthChangeListener, WeekView.EventClickListener,WeekView.EmptyViewClickListener {
    private WeekView mWeekView;
    private ArrayList<WeekViewEvent> mNewEvents;
    WeekViewEvent event;
    private WeekView.EventClickListener mEventClickListener;
    int persons[] = new int[1000];
    int i=0;
    int id=0;
    String curr;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mWeekView = (WeekView) findViewById(R.id.weekView);
        mWeekView.setMonthChangeListener(this);
        mNewEvents = new ArrayList<>();
        mWeekView.setEmptyViewClickListener(this);
        mWeekView.setOnEventClickListener(this);
    }

    @Override
    public List<WeekViewEvent> onMonthChange(int newYear, int newMonth) {
        // Populate the week view with some events.

        List<WeekViewEvent> events = new ArrayList<WeekViewEvent>();
        Calendar startTime = Calendar.getInstance();
        Calendar endTime=(Calendar)startTime.clone();

        //events
        startTime = Calendar.getInstance();
        startTime.set(Calendar.HOUR_OF_DAY, 3);
        startTime.set(Calendar.MINUTE, 0);
        startTime.set(Calendar.MONTH, Calendar.DECEMBER);
        startTime.set(Calendar.YEAR, 2015);
        endTime = (Calendar) startTime.clone();
        endTime.set(Calendar.HOUR_OF_DAY, 4);
        endTime.set(Calendar.MINUTE, 0);
        endTime.set(Calendar.MONTH, Calendar.DECEMBER);
        event = new WeekViewEvent(id++, "Available", startTime, endTime);
        event.setColor(getResources().getColor(R.color.event_color_02));
      //  Toast.makeText(MainActivity.this,""+id + "  1" , Toast.LENGTH_SHORT).show();
        events.add(event);

      startTime = Calendar.getInstance();
        startTime.set(Calendar.HOUR_OF_DAY, 6);
        startTime.set(Calendar.MINUTE, 0);
        startTime.set(Calendar.MONTH, Calendar.DECEMBER);
        startTime.set(Calendar.YEAR, 2015);
        endTime = (Calendar) startTime.clone();
        endTime.add(Calendar.HOUR_OF_DAY, 1);
        endTime.set(Calendar.MONTH, Calendar.DECEMBER);
        event = new WeekViewEvent(id++, "Available", startTime, endTime);
        event.setColor(getResources().getColor(R.color.event_color_02));
         events.add(event);

        startTime = Calendar.getInstance();
        startTime.set(Calendar.HOUR_OF_DAY, 5);
        startTime.set(Calendar.MINUTE, 0);
        startTime.set(Calendar.MONTH, Calendar.DECEMBER);
        startTime.set(Calendar.YEAR, 2015);
        startTime.set(Calendar.DATE, 16);
        endTime = (Calendar) startTime.clone();
        endTime.add(Calendar.HOUR_OF_DAY, 1);
        endTime.set(Calendar.MONTH, Calendar.DECEMBER);
        event = new WeekViewEvent(id++,"Available", startTime, endTime);
        event.setColor(getResources().getColor(R.color.event_color_03));
          events.add(event);

        startTime = Calendar.getInstance();
        startTime.set(Calendar.DAY_OF_MONTH, 16);
        startTime.set(Calendar.HOUR_OF_DAY, 2);
        startTime.set(Calendar.MINUTE, 0);
        startTime.set(Calendar.MONTH, Calendar.DECEMBER);
        startTime.set(Calendar.YEAR, 2015);
        endTime = (Calendar) startTime.clone();
        endTime.add(Calendar.HOUR_OF_DAY, 1);
        event = new WeekViewEvent(id++, "Available", startTime, endTime);
        event.setColor(getResources().getColor(R.color.event_color_04));
        events.add(event);


        ArrayList<WeekViewEvent> newEvents = getNewEvents(newYear, newMonth);
        events.addAll(newEvents);
        return events;
    }

    private ArrayList<WeekViewEvent> getNewEvents(int year, int month) {

        // Get the starting point and ending point of the given month. We need this to find the
        // events of the given month.
        Calendar startOfMonth = Calendar.getInstance();
        startOfMonth.set(Calendar.YEAR, year);
        startOfMonth.set(Calendar.MONTH, month - 1);
        startOfMonth.set(Calendar.DAY_OF_MONTH, 1);
        startOfMonth.set(Calendar.HOUR_OF_DAY, 0);
        startOfMonth.set(Calendar.MINUTE, 0);
        startOfMonth.set(Calendar.SECOND, 0);
        startOfMonth.set(Calendar.MILLISECOND, 0);
        Calendar endOfMonth = (Calendar) startOfMonth.clone();
        endOfMonth.set(Calendar.DAY_OF_MONTH, endOfMonth.getMaximum(Calendar.DAY_OF_MONTH));
        endOfMonth.set(Calendar.HOUR_OF_DAY, 23);
        endOfMonth.set(Calendar.MINUTE, 59);
        endOfMonth.set(Calendar.SECOND, 59);

        // Find the events that were added by tapping on empty view and that occurs in the given
        // time frame.
        ArrayList<WeekViewEvent> events = new ArrayList<WeekViewEvent>();
        for (WeekViewEvent event : mNewEvents) {
            if (event.getEndTime().getTimeInMillis() > startOfMonth.getTimeInMillis() &&
                    event.getStartTime().getTimeInMillis() < endOfMonth.getTimeInMillis()) {
                events.add(event);

            }
        }
        return events;
    }

   @Override
    public void onEmptyViewClicked(final Calendar time) {
       Toast.makeText(MainActivity.this,"This slot is not available" , Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onEventClick(final WeekViewEvent event, RectF eventRect) {
        Toast.makeText(MainActivity.this,""+event.getId()  , Toast.LENGTH_SHORT).show();
        Calendar time = event.getStartTime();
        String slot = String.valueOf(time);
        if (persons[((int) event.getId())] < 2 ) {

            AlertDialog.Builder alertDialogBuilder = new AlertDialog.Builder(
                    this);
            alertDialogBuilder.setTitle("Prompt");
            alertDialogBuilder
                    .setMessage("Confirm this slot for your interview?");
            alertDialogBuilder.setCancelable(false);
            alertDialogBuilder.setPositiveButton("Yes", new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int id) {
                    persons[((int) event.getId())]++;
                    Toast.makeText(MainActivity.this, ""+persons[((int) event.getId())] , Toast.LENGTH_SHORT).show();
                    mNewEvents.add(event);
                    Toast.makeText(MainActivity.this,""+event.getId()  , Toast.LENGTH_SHORT).show();
                    mWeekView.notifyDatasetChanged();
                }

            });

            alertDialogBuilder.setNegativeButton("No", new DialogInterface.OnClickListener() {
                public void onClick(DialogInterface dialog, int id) {
                    // if this button is clicked, just close
                    // the dialog box and do nothing

                    dialog.cancel();
                }
            });

            // create alert dialog
            AlertDialog alertDialog = alertDialogBuilder.create();

            // show it
            alertDialog.show();
        } else {
            if (persons[((int) event.getId())] == 2 && event.getName().equals("Available")) {
                Toast.makeText(MainActivity.this, "More than two interviewees are not allowed", Toast.LENGTH_SHORT).show();
            }
        }
    }

    private String getEventTitle(Calendar time) {
        return String.format("Event of %02d:%02d %s/%d", time.get(Calendar.HOUR_OF_DAY), time.get(Calendar.MINUTE), time.get(Calendar.MONTH)+1, time.get(Calendar.DAY_OF_MONTH));
    }
    }

**Comment by caske33** [comment]
@Swapnil-Garg16 you're still returning the "default" events three times. Only in your getNewEvents you filter... Not in your onMonthChange.

**Comment by abalta** [comment]
I know this issue closed but I have same problem here. My code belowed,

``` java
package com.pozitif.fragments;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.alamkanak.weekview.MonthLoader;
import com.alamkanak.weekview.WeekView;
import com.alamkanak.weekview.WeekViewEvent;
import com.google.android.gms.analytics.HitBuilders;
import com.pozitif.AppController;
import com.pozitif.models.ArtistModel;
import com.pozitif.onelove.R;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

/**
 * A simple {@link Fragment} subclass.
 */
public class ProgramFragment extends Fragment {

    private WeekView mWeekView;


    public static ProgramFragment newInstance() {
        ProgramFragment fragment = new ProgramFragment();
        return fragment;
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        AppController.tracker.send(new HitBuilders.EventBuilder()
                .setCategory("ui-action")
                .setAction("press")
                .setLabel("Program")
                .build());
        View rootView = inflater.inflate(R.layout.fragment_program, container, false);

        // Get a reference for the week view in the layout.
        mWeekView = (WeekView) rootView.findViewById(R.id.weekView);


        // The week view has infinite scrolling horizontally. We have to provide the events of a
        // month every time the month changes on the week view.
        mWeekView.setMonthChangeListener(new MonthLoader.MonthChangeListener() {
            @Override
            public List<? extends WeekViewEvent> onMonthChange(int newYear, int newMonth) {
                return getEvents(newYear, newMonth);
            }
        });


        Calendar startTime = Calendar.getInstance();
        startTime.set(Calendar.HOUR_OF_DAY, parseHour(LayoutFragment.getInstance().artistModelList.get(0).getStartHour()));
        startTime.set(Calendar.MINUTE, parseMin(LayoutFragment.getInstance().artistModelList.get(0).getStartHour()));
        startTime.set(Calendar.DAY_OF_MONTH, parseDay(LayoutFragment.getInstance().artistModelList.get(0).getStrDate()));
        startTime.set(Calendar.MONTH, parseMonth(LayoutFragment.getInstance().artistModelList.get(0).getStrDate()) - 1);
        startTime.set(Calendar.YEAR, parseYear(LayoutFragment.getInstance().artistModelList.get(0).getStrDate()));

        mWeekView.goToDate(startTime);

        return rootView;
    }

    private int parseHour(String hour) {
        String firstTwoStr = hour.substring(0, 2);
        return Integer.parseInt(firstTwoStr);
    }

    private int parseMin(String hour) {
        String endTwoStr = hour.substring(3, 5);
        return Integer.parseInt(endTwoStr);
    }

    private int parseDay(String date) {
        String endTwoStr = date.substring(8, 10);
        return Integer.parseInt(endTwoStr);
    }

    private int parseMonth(String date) {
        String middleTwoStr = date.substring(5, 7);
        return Integer.parseInt(middleTwoStr);
    }

    private int parseYear(String date) {
        String firstFoutStr = date.substring(0, 4);
        return Integer.parseInt(firstFoutStr);
    }

    private ArrayList<WeekViewEvent> getEvents(int newYear, int newMonth){
        Calendar startTime;
        Calendar endTime;
        List<ArtistModel> artistModels = LayoutFragment.getInstance().artistModelList;
        ArrayList<WeekViewEvent> events = new ArrayList<WeekViewEvent>();
        for (int i = 0; i < artistModels.size(); i++) {
            startTime = Calendar.getInstance();
            startTime.set(Calendar.HOUR_OF_DAY, parseHour(artistModels.get(i).getStartHour()));
            startTime.set(Calendar.MINUTE, parseMin(artistModels.get(i).getStartHour()));
            startTime.set(Calendar.DAY_OF_MONTH, parseDay(artistModels.get(i).getStrDate()));
            startTime.set(Calendar.MONTH, parseMonth(artistModels.get(i).getStrDate()) - 1);
            startTime.set(Calendar.YEAR, parseYear(artistModels.get(i).getStrDate()));
            endTime = (Calendar) startTime.clone();
            endTime.set(Calendar.HOUR_OF_DAY, parseHour(artistModels.get(i).getEndHour()));
            endTime.set(Calendar.MINUTE, parseMin(artistModels.get(i).getEndHour()));
            WeekViewEvent event = new WeekViewEvent(i, artistModels.get(i).getTitle() + " - " + artistModels.get(i).getVenueName(), startTime, endTime);
            event.setColor(ContextCompat.getColor(getActivity(), R.color.color_gray));
            events.add(event);
        }
        return events;
    }
}
```

EDIT: I solve the issue or I can get understand what the library works. There is no different year than newYear in week view. So, you must use newYear and newMonth parameter.

**Comment by jas017** [comment]
@caske33  I am facing the same issue here. 
//(you should only return the events that fall within the month that is requested, based on the arguments passed in onMonthChange)//

Could you please tell me how to do this? Any sample code?

**Comment by zkvarz** [comment]
@jas017 in your overrided onMonthChange method you just check your event's newYear and newMonth with arguments that you have in onMonthChange (newYear, newMonth).
Code goes like this:

    public List<? extends WeekViewEvent> onMonthChange(int newYear, int newMonth) {     
        List<WeekViewEvent> events = new ArrayList<>();

        for (MyEvent myEvent: eventList) {
        //TODO: this is your comparison
            if (myEvent.getMonth() == newMonth && myEvent.getYear() == newYear ) {
                Calendar startTime = myEvent.getStrartTime();
                Calendar endTime = myEvent.getEndTime();
                WeekViewEvent event = new WeekViewEvent(1, myEvent.getTitle(), startTime, endTime);
                event.setColor(ContextCompat.getColor(this, android.R.color.holo_blue_bright));
                events.add(event);
            }
        }
        return events;
    }

Hope it helps :)

**Comment by wysohn** [comment]
Wow I'm pretty sure that I read something about onMonthChange will get called three times, but I never actually realized it until visiting this thread. It's working like a charm now!

    private List<WeekViewEventWrapper> getEvents(int year, int month){
        Calendar today = Calendar.getInstance();
        today.setTime(new Date());

        if(today.get(Calendar.MONTH) != month)
            return new ArrayList<>();
        ...
    }

**Comment by RiyazShaik18** [comment]
sdasda

**Comment by JillevdW** [comment]
I know this issue is closed, but I am running into the same problem. I've read all the answers here, however I don't have the `public List<? extends WeekViewEvent> onMonthChange(int newYear, int newMonth)` function. I'm building my app with Kotlin, and my function definition looks like this `override fun onMonthChange(startDate: Calendar, endDate: Calendar): List<WeekViewDisplayable<Reservation>>`. `Reservation` is my own object.

So in short, I can't check if it's the right month, because the month is not supplied in the function. Any help is appreciated! :)

**Comment by wysohn** [comment]
@JillevdW 
Have a look into my example, then you will figure it out easily.

The startDate and endDate is Calendar, and look carefully at how I got the month of the Calender instance.

Hint:

` if(today.get(Calendar.MONTH) != month)`
